### PR TITLE
v5: Health Check & LeastLoad Strategy

### PR DESCRIPTION
**Pull request by qjebbs**
The following config explains everything:

```yaml
api:
  services:
  - RoutingService   # required for 'v2ray api bi' to work
routing:
  balancers:
  - tag: balancer
    selector:
    - all.balancer
    fallbackTag: block    # fallback to block if the balancer failed to pick one
    strategy:
      type: LeastLoad   # select least load nodes, those who have lower rtt stand deviation
      settings:
        healthCheck:
          interval: 120 # perform checks every 2 minutes (expectations, checks are random) for each outbound, default 60s
          sampling: 5   # sampling recent 5 rtts for selecting, default 5
                        # above setting evaluates the stability in last 10 minutes
          destination: http://www.google.com/gen_204
          connectivity: http://connectivitycheck.platform.hicloud.com/generate_204 # check connectivity if ping fail
          timeout: 5    # health ping timeout 5s
        costs:                          # optional. cost values are used to degrade rtt stand deviations
        - match: all.balancer.node1     # set cost=16 for specific
          value: 16
        - match: keyword                # exclude nodes by keyword
          value: 99999
        - match: cost5                  # set cost=5 if node name match cost5 (extract 5 from cost5)
        - regexp: true
          match: \.weight\d+(\.\d+)?    # matchs & set from .weight<num>, users can set weight with
                                        # outbound tags, like all.node1.weight0.5, all.node2.weight5
        tolerance: 0.1  # max acceptable failure rate
        maxRTT: 1000    # max acceptable rtt, filter away high delay nodes. defalut 0
        expected: 3     # make sure 3 nodes select (Bandwidth priority)
        baselines:      # also select nodes near above 3 according to "rtt stand deviation" baselines
        - 30
        - 40
```
More about `baselines` & `expected`:

```go
// selectLeastLoad selects nodes according to Baselines and Expected Count.
//
// The strategy always improves network response speed, not matter which mode below is configurated.
// But they can still have different priorities.
//
// 1. Bandwidth priority: no Baseline + Expected Count > 0.: selects `Expected Count` of nodes.
// (one if Expected Count <= 0)
//
// 2. Bandwidth priority advanced: Baselines + Expected Count > 0.
// Select `Expected Count` amount of nodes, and also those near them according to baselines.
// In other words, it selects according to different Baselines, until one of them matches
// the Expected Count, if no Baseline matches, Expected Count applied.
//
// 3. Speed priority: Baselines + `Expected Count <= 0`.
// go through all baselines until find selects, if not, select none. Used in combination
// with 'balancer.fallbackTag', it means: selects qualified nodes or use the fallback.
func (s *LeastLoadStrategy) selectLeastLoad(nodes []*node) []*node {
```

To inspect how `LeastLoad` strategy works to tweak settings:

```
v2ray api provides tools to manipulate V2Ray via its API.

Usage:

	v2ray api <command> [arguments]

The commands are:

        ...
        bc            balancer health check
        bi            balancer information
        bo            balancer select override
        ...

Use "v2ray help api <command>" for more information about a command.
```

```
> v2ray api bi

Balancer: selector
  - Strategy:
    random
  - Selects:
        Tag
    1   all.xxxxxxxxxx
    2   all.xxxxxxxxxx
    3   all.xxxxxxxxxx
Balancer: wrappers
  - Strategy:
    leastload, expected: 3, baselines: 30ms 40ms, max rtt: 1s, tolerance: 0.1
    health ping, interval: 2m0s, sampling: 5, timeout: 5s, destination: http://www.google.com/gen_204
  - Selects:
            RTT STD+C     RTT STD.      RTT Avg.      Hit   Cost  Tag
    1   OK  72.881354ms   72.881354ms   667.268022ms  10/10 1.00  all.balancer.xxxxxxx
    2   OK  80.283841ms   80.283841ms   773.369174ms  10/10 1.00  all.balancer.xxxxxxx
    3   OK  97.549384ms   97.549384ms   705.809272ms  9/10  1.00  all.balancer.xxxxxxx
  - Others:
            RTT STD+C     RTT STD.      RTT Avg.      Hit   Cost  Tag
    4   OK  103.248756ms  73.007896ms   951.2193ms    10/10 2.00  all.balancer.xxxxxxx
    5   OK  121.640366ms  43.006364ms   711.839948ms  10/10 8.00  all.balancer.xxxxxxx
    6   >   321.662975ms  321.662975ms  1.178536215s  10/10 1.00  all.balancer.xxxxxxx
    7   >   364.949006ms  364.949006ms  1.138114298s  10/10 1.00  all.balancer.xxxxxxx
    8   x   -             -             -             0/10  1.00  all.balancer.xxxxxxx
    9   x   -             -             -             0/10  1.00  all.balancer.xxxxxxx
    10  x   -             -             -             0/10  1.00  all.balancer.xxxxxxx
```

PS: Default strategy is Random (the original strategy), the behavior is unchanged, user doesn't need to explicitly set to it.

### CODE REVIEW

**Comment by database64128** [comment]
Can we have the `interval` in seconds? Currently it can be confusing since the `timeout` field right below it is in seconds.

An example of load balancer that does the job well is https://github.com/shadowsocks/shadowsocks-rust/blob/master/crates/shadowsocks-service/src/local/loadbalancing/server_stat.rs. It pulls TCP and UDP separately every 6 seconds, calculates the score based on different weights applied to RTT, failure rate, and stdev from the last 10 minutes.

**Comment by qjebbs** [comment]
> Can we have the `interval` in seconds? Currently it can be confusing since the `timeout` field right below it is in seconds.
> 
> An example of load balancer that does the job well is https://github.com/shadowsocks/shadowsocks-rust/blob/master/crates/shadowsocks-service/src/local/loadbalancing/server_stat.rs. It pulls TCP and UDP separately every 6 seconds, calculates the score based on different weights applied to RTT, failure rate, and latency stdev from the last 10 minutes.

Thanks for the information.

One check could take 5, 10 seconds, and even more, small seconds interval is meaningless. I take the unit as 'minutes' on purpose.

I wrote these codes without reading any other implementations (and I cannot read Rust), there of course can be better implementations, I hope someone else can continue the work since it works enough for me now:

- Improve health check logic (with RTT_WEIGHT, FAIL_WEIGHT, and more?)
- Continue to write more better strategies

**Comment by qjebbs** [comment]
@database64128 I wonder about why they just measures by `pulls TCP and UDP`, other than a real request through the outbound handler.

Suppose the outbound is setup like `frontend->IPLC->outlet` (a chain), measuring at `frontend` is unreliable, since the bottleneck and be the IPLC and outlet, or even the later nodes are failed.

**Comment by database64128** [comment]
@qjebbs shadowsocks-rust sends real requests for latency measurements. My point was, in addition to the TCP latency test with HTTP 204 pings as implemented in your PR, ss-rust also implements a UDP test by sending DNS requests. TCP and UDP scores are calculated separately, and is likely to end up selecting different servers for TCP and UDP. This helps a lot in situations where certain servers perform well with TCP traffic but poorly with UDP traffic.

**Comment by qjebbs** [comment]
This PR could possibly introduce characteristic for packet inspection in the client side(Many extreme small payload size requests in each interval of time)? I choose to keep it for personal use.

**Comment by database64128** [comment]
I'd love to see this PR merged. The feature is quite useful and extensible. And users can always choose to use it or not.

**Comment by database64128** [comment]
> This PR could possibly introduce characteristic for packet inspection in the client side(Many extreme small payload size requests in each interval of time)?

It's probably not uncommon for application protocols to have keep-alive messages that are sent at regular intervals.

**Comment by qjebbs** [comment]
@database64128  I have no expertise for this field, If the v2fly team discussed and choose to merge it, I was glad to see it happens. 😁

**Comment by kslr** [comment]
Work in this area is meaningful, I support a merge.

In the future, we don’t need a regular network test, We can get information from the route logs

**Comment by qjebbs** [comment]
> Work in this area is meaningful, I support a merge.
> 
> In the future, we don’t need a regular network test, We can get information from the route logs

Yes, it's a good plan and we can even collects speed info for balancing strategy, though the disadvantage is that, we will firstly request to a failed node, then to decide not to use it next time.

I'm not good at v2ray underlying logic, like `Link`, `Stat`, hope someone will continue on this part?

**Comment by kslr** [comment]
We can support multiple types of balancers,  And keep timing measurements.

**Comment by qjebbs** [comment]
@kslr To support speed measuring, we may need to some refactor to `Outbound.Process()` to expose `net.Conn` for statistics?

https://github.com/v2fly/v2ray-core/blob/78e01a37c7c32cda75dc62e9bf389bc3ebfcc0e8/proxy/vmess/outbound/outbound.go#L56-L57

Then:

```go
// the closest one is StatCouterConnection, but we need to implement 
// something wrapping it to calc both gloal counter and current conn speed
if stat, ok := conn.(*internet.StatCouterConnection); ok {
	dur := time.Now().Sub(start)
	speed := float64(stat.WriteCounter.Value()+stat.ReadCounter.Value()) / dur.Seconds() / 1024
	newError(
		fmt.Sprintf("dur: %s, speed: %.3f", dur, speed),
	).AtDebug().WriteToLog()
}
```

Anyway, it's not the goal of this PR

**Comment by qjebbs** [comment]
@kslr Yes, the PR supports multiple balancers with different strategies and different strategies settings. Currently we have `Random` and `LeastLoad`, but they all based on the health checker.

**Comment by qjebbs** [comment]
I think the PR is ready. The failed test is not about this PR, another randomly failed test like `TestHTTP2`?

**Comment by kslr** [comment]
```
=== RUN   TestServiceSubscribeRoutingStats
2021-01-13T02:13:15.8682750Z     command_test.go:198: rpc error: code = Unknown desc = app/router/command: Upstream closed the subscriber channel.
2021-01-13T02:13:15.8684742Z --- FAIL: TestServiceSubscribeRoutingStats (0.03s)
2021-01-13T02:13:15.8685873Z === RUN   TestSerivceTestRoute
2021-01-13T02:13:15.8689747Z --- PASS: TestSerivceTestRoute (0.01s)
2021-01-13T02:13:15.8690495Z FAIL
2021-01-13T02:13:15.8694575Z FAIL	v2ray.com/core/app/router/command	0.043s
```

Is it related?

**Comment by qjebbs** [comment]
@kslr I think no. The PR has nothing to do with `SubscribeRoutingStats`, and you can see the commit before `api bi sort output` passed the test, but the `api bi sort output` added only one func call of `sort.Slice`.

File: app/dispatcher/default.go [hunk 1]
```diff
@@ -255,6 +266,24 @@ func sniffer(ctx context.Context, cReader *cachedReader) (SniffResult, error) {
 	}
 }
 
+func (d *DefaultDispatcher) targetedDispatch(ctx context.Context, link *transport.Link, tag string) {
+	handler := d.ohm.GetHandler(tag)
+	if handler == nil {
+		newError("outbound handler [", tag, "] not exist").WriteToLog(session.ExportIDToError(ctx))
```
Comment (kslr) [inline app/dispatcher/default.go:272]: AtError ?

File: app/router/balancing.go [hunk 2]
```diff
@@ -1,44 +1,55 @@
...
+// Balancer represents a balancer
 type Balancer struct {
-	selectors []string
-	strategy  BalancingStrategy
-	ohm       outbound.Manager
+	selectors     []string
+	strategy      routing.BalancingStrategy
+	healthChecker *HealthChecker
+	ohm           outbound.Manager
+	fallbackTag   string
```
Comment (kslr) [inline app/router/balancing.go:14]: 1. selectors and fallbackTag Is there a design conflict?
2. strategy and healthChecker why are they of the same level?

File: app/router/config.go [hunk 3]
```diff
@@ -145,10 +148,56 @@ func (rr *RoutingRule) BuildCondition() (Condition, error) {
...
+			// Rounds:      int(br.HealthCheck.Rounds),
+			Timeout: time.Duration(br.HealthCheck.Timeout) * time.Second,
+		},
+		Results: nil, // nil means the check is not performed yets
+	}
+	if h.Settings.Destination == "" {
+		h.Settings.Destination = "http://www.google.com/gen_204"
+	}
+	if h.Settings.Interval == 0 {
+		h.Settings.Interval = time.Duration(15) * time.Minute
```
Comment (kslr) [inline app/router/config.go:168]: Should we avoid within 10s?

File: app/router/balancing_health.go [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+	for tag := range b.healthChecker.Results {
+		found := false
+		for _, v := range tags {
+			if tag == v {
+				found = true
+				break
+			}
+		}
+		if !found {
+			// newError("healthChecker: remove tag ", tag).AtDebug().WriteToLog()
```
Comment (kslr) [inline app/router/balancing_health.go:207]: tip

File: v2ray.go [hunk 5]
```diff
@@ -120,6 +121,10 @@ func addInboundHandlers(server *Instance, configs []*InboundHandlerConfig) error
 	return nil
 }
 
+// throttledChecker makes sure only the last call run, if many calls raised in 2 seconds
+var throttledChecker *routing.ThrottledChecker
```
Comment (kslr) [inline v2ray.go:125]: cc @xiaokangwang

File: app/router/balancing_health.go [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+	b.healthChecker.ticker = ticker
+	for {
+		_, ok := <-ticker.C
+		if !ok {
+			break
+		}
+		go func() {
+			_, err := b.CheckAll(true)
+			if err != nil {
+				newError("HealthCheckScheduler: ", err).AtWarning().WriteToLog()
```
Comment (kslr) [inline app/router/balancing_health.go:40]: Do you need a log-prefix

File: app/router/balancing_health.go [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+// it validates tags and checks those only in current balancer
+func (b *Balancer) Check(tags []string, distributed bool) error {
+	ts, err := b.getCheckTags(tags)
+	if err != nil {
+		return err
+	}
+	if len(ts) == 0 {
+		return nil
+	}
+	newError("HealthCheck: Perform one-time check for tags ", ts).AtInfo().WriteToLog()
```
Comment (kslr) [inline app/router/balancing_health.go:74]: Do you need a log-prefix

( only these two)

File: app/dispatcher/default.go (hunk 1 above)
Comment (qjebbs) [reply app/dispatcher/default.go:272]: My mistake

File: app/router/balancing.go (hunk 2 above)
Comment (qjebbs) [reply app/router/balancing.go:14]: 1. No. fallbackTag is a option to change the default fallback behavior (to default outbound, it's not predictable), if not set, it the previous behavior
2. healthChecker is not part of strategy. Imagine that we add different strategies based on the same healthChecker

File: app/router/config.go (hunk 3 above)
Comment (qjebbs) [reply app/router/config.go:168]: I agree.

File: app/router/balancing_health.go (hunk 4 above)
Comment (qjebbs) [reply app/router/balancing_health.go:207]: Sorry, I don't understand

Comment (kslr) [reply app/router/balancing_health.go:207]: = = forgot to write, do you still need it here

File: v2ray.go (hunk 5 above)
Comment (qjebbs) [reply v2ray.go:125]: this is designed to avoid repeatedly lock & unlock of health check results.

File: app/router/balancing_health.go (hunk 7 above)
Comment (qjebbs) [reply app/router/balancing_health.go:74]: By what means I can add a log-prefix?

File: app/router/balancing_health.go (hunk 4 above)
Comment (qjebbs) [reply app/router/balancing_health.go:207]: log-prefix? yes, I can change it though it's commented. But by what means I can add a log-prefix? Any tips?

**Review by kslr — APPROVED** [review]

File: v2ray.go (hunk 5 above)
Comment (qjebbs) [reply v2ray.go:125]: 我打算把这部分增改触发检查的代码移除。

原先的场景是没有把检查分摊开来，如果有outbound添加或更新，立即检查就可以尽快用上它们。现在分摊开来后，假设检查间隔是5分钟，那么这里触发的检查完成的期望值是2.5分钟。等2.5跟等5没有本质差别。删除掉后：
- 可以降低代码复杂度
- 用户在添加更新outbound后，可以手工强制更新 `v2ray api bc`

File: app/router/balancing_health.go (hunk 4 above)
Comment (qjebbs) [reply app/router/balancing_health.go]: comment removed

File: v2ray.go (hunk 5 above)
Comment (qjebbs) [reply v2ray.go]: 上面说到，分摊后立即检查没用了，这是坏处。好处是rounds有意义了：
- 原先如果测试5回合，那么几乎是在同一时间测的，rtt相近，没什么意义
- 现在分摊开来，相当于这5回合分布在分钟内测的，可以衡量稳定性

**Comment by qjebbs** [comment]
@kslr 不好意思，你审阅后随着使用体验，又改了好多，这次变化比较大：

1. 原来是根据平均rtt选择，往往选择了地理位置近这种近的节点，并不一定代表他们负载少。
2. 现在根据多次rtt的方差选择，方差小的说明线路稳定：包括节点本身负载少、线路干扰少。实测效果更好，在晚高峰能选出支持4K的节点
3. 原来的配置方式不好，现在改为更直观地针对每个节点的测试间隔（期望值，实际上是随机的），比如每2分钟一次
4. 原来回合配置不再有意义，改为采样数量配置。假设配置为5，意思是采样最近5次的结果，结合上述2分钟一次，即可评估最近10分钟内的稳定性。
5. 因为检查结果实时提交，采样最近5次是无缝的，不会像之前一样，这5次与下5次分开独立评估。
6. 总地来说，评估的周期、数量、间隔都可以调整
7. 正如策略名称暗示的那样，这个策略聚焦的是低负载（表现稳定），而非低延迟。但还是增加了一个maxRTT设置，允许把过高延迟的节点剃除

> 相比之下，上一版本（未根据标准差）更像是一个寻找低延迟的策略。

**Comment by LazyZhu** [comment]
@qjebbs 
balancers 不支持多个组么? 设置多个组也只会对配置文件中排序第一个进行测试.
    
        "balancers": [
            {
                "tag": "as-out",
                "selector": [
                    "server1a",
                    "server1b",
                    "server1c"
                ],
                "strategy": {
                    "type": "LeastLoad",
                    "settings": {
                        "healthCheck": {
                            "destination": "http://detectportal.firefox.com/success.txt",
                            "timeout": 5,
                            "interval": 60,
                            "sampling": 10
                        },
                        "maxRTT": 500,
                        "expected": 3,
                        "baselines": [
                            20,
                            30,
                            50,
                            100,
                            200
                        ]
                    }
                }
            },
            {
                "tag": "na-out",
                "selector": [
                    "server2a",
                    "server2b",
                    "server2c"
                ],
                "strategy": {
                    "type": "LeastLoad",
                    "settings": {
                        "healthCheck": {
                            "destination": "http://detectportal.firefox.com/success.txt",
                            "timeout": 5,
                            "interval": 60,
                            "sampling": 10
                        },
                        "maxRTT": 1200,
                        "expected": 0,
                        "baselines": []
                    }
                }
            },
            {
                "tag": "eu-out",
                "selector": [
                    "server3a",
                    "server3b",
                    "server3c"
                ],
                "strategy": {
                    "type": "LeastLoad",
                    "settings": {
                        "healthCheck": {
                            "destination": "http://detectportal.firefox.com/success.txt",
                            "timeout": 5,
                            "interval": 60,
                            "sampling": 10
                        },
                        "maxRTT": 1500,
                        "expected": 3,
                        "baselines": [
                            50,
                            100,
                            150,
                            200,
                            300
                        ]
                    }
                }
            }
        ]

**Comment by qjebbs** [comment]
@LazyZhu 有个大BUG，用最新的应该修复了

@kslr 新增了 cost加权、failure tolerance、手工覆盖均衡器、网络连通性检查。想做的都做了，没做的做不了或者不想做了😂

> 没做的是指手工覆盖时，暂停检查：导致数据竞争，能解决、但没有简单优雅的方法可以解决

请帮忙重点看下racing的情况，我已经跌了好几个坑了。。

**Comment by qjebbs** [comment]
好了，收工。

> 回头看核心策略部分不过300行代码（原以为事情很简单，检查再挑嘛），给它配套的却有3000行，弄到想吐。  
> 现在从ping、采集、统计、挑选、输出统计、手工控制都齐了，未来要写新策略大概率能在几百行的规模上搞定，包括不限于：
> - 按权重将udp和tcp一起考察
> - 按权重将rtt和失败率一起考查
> - ip hash 防止出口频繁变化
> - ...
> 
> 等有缘人吧。

**Comment by kslr** [comment]
Great! 只差写一下文档了

p.s 很有可能，不过这个功能作者已经弃坑了，如果未来一段时间没人接手，我可能会选择删掉，换一种简单的方式实现 （ 指找一个库换上

**Comment by kslr** [comment]
:fireworks:

**Comment by acc557** [comment]
@kslr @qjebbs 

Is this balancer battery-firendly?  
If I set interval to 2147483648 , will it perform health check when the failure rate of current outbound is high?

**Comment by iusearch** [comment]
这个合到v5分支了，但是现在的v5 release只有LeastLoad的添加，那几个api command都没有，想问问怎么回事？

**Pull request merged**